### DISCUSSION TITLE
what are global Util do in dashboard-nvim config?

### DISCUSSION THREAD

**Discussion by cStralpt** [LazyVim/LazyVim#1748] [Q&A] [answered]
i got this error
``
E5108: Error executing lua: [string "Util.telescope.config_files()"]:1: attempt to index global 'Util' (a nil value)
stack traceback:
	[string "Util.telescope.config_files()"]:1: in function 'dump'
	...e/nvim/lazy/dashboard-nvim/lua/dashboard/theme/hyper.lua:41: in function <...e/nvim/lazy/dashboard-nvim/lua/dashboard/theme/hyper.lua:35>
``
when pressing c to open nvim config

**Comment by folke** [comment, marked as answer]
You probably copied the code? Add `local Util = require("lazyvim.util")`

**Reply by rafo** [reply]
I believe the issue stems from the template "Options" at https://www.lazyvim.org/plugins/ui#dashboard-nvim. It lacks `local Util = require("lazyvim.util")`.

**Reply by dpetka2001** [reply]
Lazyvim website only contains a snippet of the code, which mentions the Lazyvim specific fragment of the plugin spec. Since, the code for the website is autogenerated, I guess that it only parses the specific spec of a plugin. If `local Util = require("lazyvim.util")` is used throughout that file by different plugin specs, you can't expect to be defined in every plugin spec. That's why it's only defined once, at the top of the file, so that all specs can benefit from this and I guess that's the reason it fails to be parsed. You can't expect all the code of the Lazyvim codebase to be included on the website. Sometimes, you have to go through the github repo to find answers and if someone searches for `Util.telescope`, I'm sure he can dig throughout the results to find what's missing.

**Reply by folke** [reply]
Indeed. That's there as a reference only

**Reply by rafo** [reply]
@dpetka2001 @folke I can understand that well!

However, this is a documentation issue and should be thoroughly mentioned at least on https://www.lazyvim.org/configuration/plugins, as otherwise, some NeoVim beginners might be left behind.

I myself use the variant with multiple plugin files under "plugins" instead of just one file because I find it more organized. However, the documentation gives the impression that the code in the "Options" tab should simply be copied into a file for the plugin to work.

For example, if you look at https://www.lazyvim.org/configuration/plugins or in the examples at https://www.lazyvim.org/configuration/examples, you can see code examples like this:

```
return {
  "nvim-telescope/telescope.nvim",
  ...
```
In the documentation for the other plugins, this essential code snippet for the file-per-plugin variant is missing in the "Options" tab.

Perhaps you could add the following (commented out) to the "Options" section (looks horrible IMO):

```
-- Uncomment lines to manage plugins with a file each. See https://www.lazyvim.org/configuration/plugins
-- return {
--      "nvimdev/dashboard-nvim",
...
        opts = function()
            -- Uncomment the lines below to manage plugins with a file each. See https://www.lazyvim.org
            -- local Util = require("lazyvim.util")
...
-- Uncomment the lines below to manage plugins with a file each. See https://www.lazyvim.org
-- }
```

if you don't want to add a third tab "Options Single Plugin Files" ;D ? It would be better to write this in detail in the documentation. I would like to help!

For beginners, this is very confusing and not easy to understand. And Lazyvim is more intended for Vim beginners, right?

One more thing: I'm not a LUA expert (and honestly find the ambiguity of LUA for configurations **terrible**), but how does the definition of `local Util = require("lazyvim.util")` and similar work when maintaining all plugins in a single file? Are they inherited from the "Full Spec"?

**Reply by dpetka2001** [reply]
From [Plugins](https://www.lazyvim.org/configuration/plugins) section of the docs.
> Configuring LazyVim plugins is exactly the same as using lazy.nvim to build a config from scratch.
For the full plugin spec documentation please check the lazy.nvim [readme](https://github.com/folke/lazy.nvim).

So, I would think that the user would also read `lazy.nvim`'s docs to understand how things work. The same goes for the rest of the plugins. You can't expect Lazyvim docs to include all documentation of every single plugin it provides. The user should refer to the corresponding docs to read and understand things. Also in [Adding Plugins](https://www.lazyvim.org/configuration/plugins#-adding-plugins) there are examples mentioning how to construct the spec of a plugin. So, the rest of the docs assumes that the user has read this part and is not necessary to include the `return` part in every single spec snippet. In a single file you would only have to include `local Util = require("lazyvim.util")` at the top of the file once and since that would be `require`d through the parsing phase of `lazy.nvim`, there would be no problems when using that specific `Util` in the rest of your plugins' specs.

**Reply by rafo** [reply]
> So, I would think that the user would also read lazy.nvim's docs to understand how things work. **The same goes for the rest of the plugins**.

What would be the point of LazyVim then? That's exactly what we want to save users from with the "buildin Plugins".

> You can't expect Lazyvim docs to include all documentation of every single plugin it provides.

I didn't say that, and I didn't demand that either (I hope I dont sound demanding at all! I am extremly thankful for LazyVim). But from the perspective of an inexperienced user, I will repeat myself:

> ... the documentation gives the impression that the code in the "Options" tab should simply be copied into a file for the plugin to work.

The point is that everyone configures LazyVim as they envision it, and LazyVim does this job wonderfully and keeps getting better (:LazyExtras). The plugin configurations **are documented more extensively than necessary**. If "Full Spec" and "Options" were not documented, everyone would automatically start searching in the code and on the plugin websites (but thats not the goal of something with Lazy in the name). The danger is that errors that are difficult to solve can occur because it is not the case.

Dont get me wrong: Only a little bit is missing for the configuration in individual files per plugin. We have already written more in this conversation than would be necessary for that. ;D

Just to clarify, consider the perspective of a NeoVim newcomer on your last sentences:

> So, the rest of the docs assume that the user has read this part and it is not necessary to include the return part in every single spec snippet.

Is this for a single configuration file or multiple? Ah, maybe the next sentence will clarify:

> In a single file, you would only have to include local Util = require("lazyvim.util") at the top of the file once, and since that would be required through the parsing phase of lazy.nvim, there would be no problems when using that specific Util in the rest of your plugins' specs.

But how can I manage it the linux way for multiple config files like its mentioned in the docs? At the beginning of each single file? Or just one since all are somehow parsed? Can I omit return{} when using "Util"?

Documentation is annoying, I know. But in the long run, it saves all developers a lot of work (and I see some questions in the discussions and on r/vim and r/neovim that wouldn't even be asked if the documentation were more comprehensive).

**Reply by dpetka2001** [reply]
>> So, I would think that the user would also read lazy.nvim's docs to understand how things work. The same goes for the rest of the plugins.

>What would be the point of LazyVim then? That's exactly what we want to save users from with the "buildin Plugins".

Is this a serious answer? What would be the point of LazyVim if the users were expected to read the docs? I'm not going to get into such an argument. 

>> So, the rest of the docs assume that the user has read this part and it is not necessary to include the return part in every single spec snippet.

> Is this for a single configuration file or multiple? Ah, maybe the next sentence will clarify:

This is for every file that you want to **return** plugin specs from within. And this is getting into `lazy.nvim` territory already, which is not the responsibility of Lazyvim to teach the user about, except from the fact that it links to `lazy.nvim`'s docs for the user who might be interested.

> Can I omit return{} when using "Util"?

`return` and `Util` are 2 entirely different things. First is used to return spec from a single or multiple plugins to `lazy.nvim`. It's the same as doing 
```lua
local M = {
  -- your spec here
  ...
}
return M
```
in a file. Each file returns a table to `lazy.nvim` and inside that table you can either have a single plugin spec or multiple plugins' specs, each one in its own sub-table inside the outer table that will get returned. Again, this is `lazy.nvim` territory and a user should read the docs if he would like to understand in further depth.
The latter `Util` is specifically for requiring a lazyvim module that offers some utility modules with functions. If you don't want to make use of them, you don't have to use them. If someone would like to use them, then he should read the code of those modules to see what he can use. Again, this is out of Lazyvim's website docs scope.
> (and I see some questions in the discussions and on r/vim and r/neovim that wouldn't even be asked if the documentation were more comprehensive)

Most of the questions I see are due to lack of user's effort to read through the docs. I've seen very few plugins with non-comprehensive documentation. The docs are there for a reason and in my opinion it's an insult to the plugins' authors, after the effort and work they put into this to write them, to say 
> What would be the point of LazyVim then? That's exactly what we want to save users from with the "buildin Plugins".

**Reply by rafo** [reply]
> Is this a serious answer? What would be the point of LazyVim if the users were expected to read the docs? I'm not going to get into such an argument.

That was for the **bold part** in my comment: **The same goes for the rest of the plugins.** I don't believe LazyVim intends for a user to not only understand the LazyVim configuration but also understand the Lazy plugin manager ***and then also read and understand every plugin configuration on the plugin pages... 

Please, just take a moment to relax!

> This is for every file that you want to return plugin specs from within.

You don't have to explain this and the "Util" part, since I wrote: **_"Just to clarify, **consider the perspective of a NeoVim newcomer** on your last sentences:"_**

> And this is getting into lazy.nvim territory already, which is not the responsibility of Lazyvim to teach the user about, except for the fact that it links to lazy.nvim's docs for the user who might be interested.

Yep, that's true and was not up for discussion, except that LazyVim has already documented a lot of lazy.nvim configuration...

Sorry, you are missing my point, as I'm not arguing about that but about

> The plugin configurations are documented more extensively than necessary. If "Full Spec" and "Options" were not documented, everyone would automatically start searching in the code and on the plugin websites (but that's not the goal of something with Lazy in the name). The danger is that errors that are difficult to solve can occur because it is not the case.

I repeat myself **_Don't get me wrong: Only a little bit is missing for the configuration in individual files per plugin. We have already written more in this conversation than would be necessary for that._** ;D

I don't know what your role is here in combination with @folke, but if it is pure gatekeeping, then I am now informed. Maybe there's a lack of understanding for certain nuances in language (no offence intended)? I just wanted to help keep the number and height of barriers for the "average" user as low as possible. If you think the plugin authors are offended when their docs are not read: if a user copies all the snippets into a file, they only have to do that for the few plugins whose defaults are not sufficient.

> Most of the questions I see are due to lack of user's effort to read through the docs.

**That's** an insult.

I would have preferred to invest the time in improving the documentation (and most developers are thankful for that), but I think that would be a waste of time if your attitude matters.

**Reply by dpetka2001** [reply]
No relationship with @folke at all. This is just me purely speaking as a user. I've not even contributed that much to the project. Others have made more meaningful contributions. The written text is sometimes difficult to comprehend, when you're not familiar with the person you're interacting, especially if you don't speak to each other in person, as is the case with us. Therefore, I apologize for not understanding correctly what you really wanted to point out. I honestly thought you were implying that users should not read the docs and also not be encouraged to do so. If you really have the will to contribute to the docs to make them even clearer, then please go ahead and submit a PR with the changes you deem necessary. Folke will be the one reviewing. As for the gatekeeping part, I honestly have no intention of doing so. I'm always trying to help as much as I can with user configurations (at least for the parts that I'm able to help with, there's also a lot that I can't), not only here but in Reddit as well. I don't have anything to gain. I do it, because I like it and I also believe that trying to do so, will also help me deepen my knowledge even further. I don't believe behavior like this goes hand in hand with gatekeeping in my humble opinion.
I hope I've made things clearer from my part as you've made yours for me with your last post.

**Comment by abeldekat** [comment]
Hello @rafo,

Out of interest, I read the conversation above and noticed the line quoted below in this [comment](https://github.com/LazyVim/LazyVim/discussions/1748#discussioncomment-7290795):

> ... And Lazyvim is more intended for Vim beginners, right?

In my opinion, this is not correct. When reading the [documentation](https://www.lazyvim.org/), I do not see any text supporting that assumption.

From the headlines of the documentation:
> LazyVim is a Neovim setup powered by [💤 lazy.nvim](https://github.com/folke/lazy.nvim) to make it easy to customize and extend your config.

I think many experienced users consider `LazyVim` a core part of their *own* config, and have a solid understanding of what's happening under the hood.
